va/va.go

```diff
@@ -734,6 +747,14 @@ func (va *ValidationAuthorityImpl) PerformValidation(ctx context.Context, req *v
 	// singular problem, because the remote VAs have already audit-logged their
 	// own validation records, and it's not helpful to present multiple large
 	// errors to the end user.
-	prob = va.performRemoteValidation(ctx, req)
+	op := func(ctx context.Context, remoteva RemoteVA, req proto.Message) (remoteResult, error) {
```

Comment:
I think this problem is called covariance vs contravariance and it comes up in Rust more often because of lifetimes (and kinda broke my brain): https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)

Basically what you'd want here is to be able to say "`performRemoteValidation` can take any function that takes these parameters _or anything more specific_". But Go avoids such type complexities. Rust will make you happy!

Reply:
Oh geez I haven't thought about this since I took PLs junior year.

I think the issue here is more fundamental than covariance vs contravariance: Go treats the func type as a concrete type (as opposed to an interface type), which therefore can't be satisfied by anything except a variable with that exact type. The fact that some of the function's arguments are interface types is no different from a struct having some member fields that are interface types.

But I think that even if Go did allow this, _then_ you're totally right that I'd run into contravariance problems. The function operator is contravariant in its inputs, meaning that you can't assign a function that is strict about what arguments it accepts to a variable whose type is a function that is loose about what it accepts -- the concrete function can't handle the whole input space. So substituting `vapb.PerformValidationRequest` for `proto.Message` in the parameters list would be unacceptable, even in a language that allowed subtyping functions.

*sigh* thanks for bringing me back